### PR TITLE
Disable -Ydelambdafy:method for specialized FunctionN

### PR DESCRIPTION
**Pull request by retronym**
The Delambdafy phase generates its `FunctionN` subclasses after
the specialization phase. As such, `((x: Int) => x).apply(42)` incurs
boxing.

This commit falls back to the `-Ydelambdafy:inline` in this case.

This is done by running the specialization type map over the
type of the function, and seeing if anything changes. To make this
work robustly, we first need to ensure that the specialization info
transformer has processed all the function types.

This is not a fundamental limitation; we could in principle generate
the specialized code.

A followup change will use `-Ydelambdafy:method` as the basis for
invokedymnamic lambdas. As part of that stream of
work, we will synthesize specialization-aware lambdas, and remove
the fallback to `-Ydelambdafy:inline`.

Review by @gkossakowski

I have updated some tests that indend to test the delambdafy transform
to avoid use of specialized function types.

### CODE REVIEW

**Comment by gkossakowski** [comment]
It looks like this failure is legit:

```
java.lang.AssertionError: assertion failed: [[syntax trees at end of                delambdafy]] // newSource1.scala
package o {
  package a {
    class C extends Object {
      def hihi(): List = immutable.this.List.apply(scala.this.Predef.wrapIntArray(Array[Int]{1, 2})).map({
  (new <$anon: Function1>(C.this): Function1)
}, immutable.this.List.canBuildFrom()).$asInstanceOf[List]();
      def <init>(): o.a.C = {
        C.super.<init>();
        ()
      }
    };
    @SerialVersionUID(value = 0) final <synthetic> class anonfun$hihi$1 extends scala.runtime.AbstractFunction1$mcII$sp with Serializable {
      final def apply(x$1: Int): Int = anonfun$hihi$1.this.apply$mcII$sp(x$1);
      <specialized> def apply$mcII$sp(x$1: Int): Int = x$1.*(2);
      final <bridge> <artifact> def apply(v1: Object): Object = scala.Int.box(anonfun$hihi$1.this.apply(scala.Int.unbox(v1)));
      def <init>($outer: o.a.C): <$anon: Function1> = {
        anonfun$hihi$1.super.<init>();
        ()
      }
    }
  };
  package a {
    object `package` extends Object {
      def f(): Int = 1;
      def <init>(): o.a.package.type = {
        `package`.super.<init>();
        ()
      }
    }
  }
}


    at Test$.show(t9097.scala:32)
```

**Comment by retronym** [comment]
That was a newly added test in the category of test that "I updated some tests that intend to test the delambdafy transform to avoid use of specialized function types."

I've just changed it to use a non-specialized function so that delambdafy is still used.

**Comment by retronym** [comment]
Just tweaked another of a test that intends to exercise `-Ydelambdafy:method`, but was using a `(Int => Int)` function whose encoding differs after this PR.

**Comment by gkossakowski** [comment]
/rebuild

I'm not sure why those tests are failing.

File: src/compiler/scala/tools/nsc/transform/UnCurry.scala [hunk 1]
```diff
@@ -70,6 +70,14 @@ abstract class UnCurry extends InfoTransform
     private val noApply           = mutable.HashSet[Tree]()
     private val newMembers        = mutable.Map[Symbol, mutable.Buffer[Tree]]()
 
+    private lazy val forceSpecializationInfoTransformOfFunctionN: Unit = {
+      if (currentRun.specializePhase != NoPhase) {
```
Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/transform/UnCurry.scala]: Why is this needed?

Comment (retronym) [inline src/compiler/scala/tools/nsc/transform/UnCurry.scala]: I don't recall. Maybe it is for custom pipelines like ScalaJS.

Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/transform/UnCurry.scala]: If it's not clear why it's needed, should we remove it? This kind of code is hard to remove later on because everybody fears some hard to predict breakage.

**Comment by retronym** [comment]
@gkossakowski By removing that condition, I found that a test case using `-Ystop-after` uncurry failed. That must have been what prompted by to add this code in the first place. I have reinstated it with a supporting comment.

**Comment by retronym** [comment]
@adriaanm / @lrytz could I ask for a review of this one?

**Comment by adriaanm** [comment]
LGTM -- Merging to unlock next steps. There's a small typo btw. Left as exercise for the submitter :)

**Comment by adriaanm** [comment]
I was just about to hit the merge button, @retronym -- meant to do sooner, sorry.

**Comment by adriaanm** [comment]
PS: canUseDelam_b_dafyMethod
